protongraph/protongraph
# Changes to get a particular generator working in an example project

@HungryProton @fire 

Sorry that this is a bit messy, but I'm not very experienced in contributing to open source projects.

(Note: Pull request against the develop branch, not master.)

(Further Note: This was developed against Godot 3.4rc1.)

### Description

Enabling changes to allow fence.tpgn [in this project](https://github.com/token-cjg/TestProtongraphProject) to generate a fence along a curve when invoking this application, also when compiled locally.

A few things have been modified in this change:

* Alters the way resources are handled, so that resources are passed to the generator from the client by reference (as resource_references) before being passed back to the client for rendering.
* Some changes to the public_api and in a number of other places to get things working.
* Comments out some C++ code, this was required in order to be able to build for OSX.  See the issue here https://github.com/protongraph/protongraph/issues/100 for more context.

Note that these changes were worked on in parallel with changes to the legacy ConceptGraph code and the Sync-Godot addon manually added to this example project: https://github.com/token-cjg/TestProtongraphProject.

### Testing Instructions
* Run ./godot.3.4rc1 to run Protongraph with these changes mounted
* Open the example project https://github.com/token-cjg/TestProtongraphProject in Godot 3.4rc1.
* Ensure that you activate the ProtonGraph Sync plugin within "assetlib->addons" in Godot 3.4rc1.
* Open scenes/TestProtongraph.tscn.
* Click on the "Fence" node in the scenetree
* Make sure that the example project is connected to Protongraph via sync-godot (you should see a green check in the Node details UI)
* Click "Rebuild" in the Node details UI for the Fence node in the TestProtongraph scene within the example project.
* Wait a fraction of a second for the data in the Input subnode of the Fence object to be compiled by Sync-Godot, serialized, sent to Protongraph, deserialized, worked on, serialized and sent back via the WebSocket connection to the waiting Sync-Godot module in the Test project, deserialized, interpreted, and rendered.
* You should see a fence generated as a sequence of output nodes.

### References
* Youtube demonstration [here](https://www.youtube.com/watch?v=xN11INQlro8&ab_channel=ChrisGoddard)
* Example project that is required in order for this to work: https://github.com/token-cjg/TestProtongraphProject

### Demo (click on the image to view)

[![Kicking the tires of Protongraph](https://img.youtube.com/vi/xN11INQlro8/0.jpg)](https://www.youtube.com/watch?v=xN11INQlro8&ab_channel=ChrisGoddard)
